lib/src/modules/task.cc

```diff
@@ -0,0 +1,320 @@
+    wrapper_input.set<std::string>("stdout", (results_dir / "stdout").string());
+    wrapper_input.set<std::string>("stderr", (results_dir / "stderr").string());
+    wrapper_input.set<std::string>("exitcode", (results_dir / "exitcode").string());
+
+    auto exec = lth_exec::execute(
+        wrapper_executable_,
+        {},
+        wrapper_input.toString(),
+        environment,
+        [results_dir](size_t pid) {
+            auto pid_file = (results_dir / "pid").string();
+            lth_file::atomic_write_to_file(std::to_string(pid) + "\n", pid_file);
+        },  // pid callback
+        0,  // timeout
+        leatherman::util::option_set<lth_exec::execution_options> {
+            lth_exec::execution_options::thread_safe,
+            lth_exec::execution_options::create_detached_process,
+            lth_exec::execution_options::merge_environment,
+            lth_exec::execution_options::inherit_locale });  // options
+
+    if (exec.exit_code == TASK_WRAPPER_FILE_ERROR_EC) {
+        // This is unexpected. The output of the task will not be
+        // available for future transaction status requests; we cannot
+        // provide a reliable ActionResponse.
+        std::string empty_label { lth_loc::translate("(empty)") };
+        LOG_WARNING("The task wrapper process failed to write output on file for the {1}; "
+                    "stdout: {2}; stderr: {3}",
+                    request.prettyLabel(),
+                    (exec.output.empty() ? empty_label : exec.output),
+                    (exec.error.empty() ? empty_label : exec.error));
+        throw Module::ProcessingError {
+            lth_loc::translate("failed to write output on file") };
+    }
+
+    // Stdout / stderr output should be on file; read it
+    response.output = storage_.getOutput(request.transactionId(), exec.exit_code);
+    processOutputAndUpdateMetadata(response);
+}
+
+ActionResponse Task::callAction(const ActionRequest& request)
+{
+    auto task_execution_params = request.params();
+    auto task_input_format = task_execution_params.includes("inputFormat") ?
+        task_execution_params.get<std::string>("inputFormat") : std::string{""};
+    std::map<std::string, std::string> task_environment;
+    std::string task_input;
+
+    bool has_input = false;
+    if (task_input_format.empty() || task_input_format == "stdin:json") {
+        task_input = task_execution_params.get<lth_jc::JsonContainer>("input").toString();
+        has_input = true;
+    }
+
+    if (task_input_format.empty() || task_input_format == "environment") {
+        addParametersToEnvironment(task_execution_params.get<lth_jc::JsonContainer>("input"), task_environment);
+        has_input = true;
+    }
+
+    if (!has_input) {
+        throw Module::ProcessingError {
+            lth_loc::format("unsupported task input format: {1}", task_input_format) };
+    }
+
+    auto task_command = getTaskCommand(
+        getCachedTaskFile(task_cache_dir_,
```

Comment:
I think I have a simple fix for this, but would prefer to handle that in a separate PR.